Add the first CPP test (using Gtest)

icyflame opened the pull request:
- Test added using the GTest package
- I could not include the `bottomcam.cpp` file because it already had a `main()` function. We may have to restructure all our code, so that there is an API with which the `main()` function interacts, rather than writing all the code in the `main()` function.
- You can run this test using `make test`. The tests are just sanity tests for now, so, they will pass. 
- Full command output is available here: https://gist.github.com/icyflame/90c5f1f84b0ea65105d9

icyflame committed:
Add the first CPP test (using Gtest)

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame committed:
Add the Camera class

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame committed:
Add the main.cpp file, and create an executable for it.

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

hargup commented on vision_stack/camera/bottomcam/test/test_cam_opens.cpp:
```diff
@@ -0,0 +1,23 @@
+*/
+// Run all the tests that were declared with TEST()
+int main(int argc, char **argv){
+  testing::InitGoogleTest(&argc, argv);
```
Sorry for jumping in, but what are `GoogleTest`? We TDD works for AUV we can definitely transfer these lessons to AGV.

icyflame replied:
@hargup Thanks for jumping in!

[Google Test](http://wiki.ros.org/rostest/Writing#With_gtest) is the CPP side of [rostest](http://wiki.ros.org/rostest).  
I am basically restructuring this package so that we can separate the API, and the `main()` function. [Previously, we had code](https://github.com/auviitkgp/kraken_3.0/blob/d7c7d1cdcc6deacb10f62c50df6a880b5c25db1b/vision_stack/camera/bottomcam/src/bottomcam.cpp#L1) that had the API and the `main()` function baked into the same file, and thus, when I tried to bring it in the test CPP file, it threw an error. (2 `main` functions.)  
You can see the [file `test/*.cpp`](https://github.com/icyflame/kraken_3.0/blob/b4dcd710512c49a59483e8179f89f36a381a1181/vision_stack/camera/bottomcam/test/test_cam_opens.cpp) which is the test file. I plan to include the `*.h` API there, and then, write tests for it. The tests work fine for now, but I have not written any meaningful tests yet.  
(Currently, there are only some sanity tests there. 1 == 1 and 10 > 0 :smile: )

---

cc @nevinvalsaraj @prudhvid

hargup replied:
cc @abinashmeher999, @shivamvats, @krishna95

icyflame committed:
Function definitions removed from class definition

- Add to src/bottomcam.cpp

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame committed:
Add meaningful tests

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame committed:
Remove unnecessary code.

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame commented:
@nevinvalsaraj I have added meaningful tests in the `test/*.cpp` file. This test will open the camera and retrieve an image from there, and see if that image is empty or not.
***
cc @prudhvid @hargup

icyflame committed:
[format-code] Inside the bottomcam test suite.

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame committed:
Create executable inside bin, instead of bin/test

- The directory bin/test was not automatically created by Make and
  resulted in the test not running at all.

Attn: @kalyan-kumar

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame commented:
Tested on @kalyan-kumar 's machine. Works without having to install any extra packages. (`libgtest-dev` is a requirement. `rostest` is bundled with the complete ROS installation.)

LGTM.

***

cc @prudhvid @nevinvalsaraj

icyflame commented:
Checked on @aniket11bh 's machine too, on a brand new installation of ROS.

LGTM.

Will merge tomorrow night. If anyone has anything else to suggest, please do so by then.

***

cc @prudhvid @nevinvalsaraj @VishnuDuttSharma @aksonuabhay

prudhvid commented on vision_stack/camera/bottomcam/include/bottomcam/bottomcam.h:
```diff
@@ -0,0 +1,34 @@
+public:
+    Camera(int camNo=0);
+    cv::VideoCapture cam;
+    int cameraNo;
```
When you provide the get/set methods, I think this variable access should be made private.

icyflame replied:
@prudhvid Do you mean the `cv::VideoCapture` variable?

prudhvid replied:
@icyflame cameraNo because you have getCameraNo and setCameraNo

icyflame replied:
@prudhvid Yeah, I will do it.

prudhvid commented on vision_stack/camera/bottomcam/src/main.cpp:
```diff
@@ -0,0 +1,79 @@
+
+            // cv::imshow(_image.image);
+
+            _publishImage = _image.toImageMsg();
```
I think we should shift to compressed images from now for most of the image transfer fills up the buffer unnessarily.
http://wiki.ros.org/compressed_image_transport

icyflame replied:
@prudhvid The earlier approach was not using Compressed Image Transport.

And this PR is only to merge the test, we can move to that in a separate PR. What do you think?

---

cc @nevinvalsaraj

prudhvid replied:
yeah. thats fine

prudhvid commented on vision_stack/camera/bottomcam/src/main.cpp:
```diff
@@ -0,0 +1,79 @@
+    {
+        int _cameraNo = atoi(argv[1]);
+        camObj.setCameraNo(_cameraNo);
+        _this_cam.open(_cameraNo);
```
I think we should also handle the case when this function call fails. And then return an error message specifying this.

prudhvid commented on vision_stack/camera/bottomcam/src/main.cpp:
```diff
@@ -0,0 +1,79 @@
+        {
+            std::cout << "Camera " << camMsg << " opened successfully.";
+            camObj.setCameraNo(camMsg);
+        }
```
There can be an else here if the open function failed

icyflame replied:
Added the `else` statement.

icyflame committed:
Change variable levels, handle open camera errors.

This is a combination of 2 commits.
The first commit's message is:

Make cameraNo variable private

Attn: @prudhvid

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

This is the 2nd commit message:

Add else statements to both camera open statements

Signed-off-by: Siddharth Kannan <kannan.siddharth12@gmail.com>

icyflame commented:
@prudhvid @nevinvalsaraj Added the things that were requested at the earlier diff. Please review now.

nevinvalsaraj commented:
@icyflame I see that the updated PR doesn't have major changes. Squash the minor commits together before merge.

icyflame commented:
@nevinvalsaraj Will do. I will squash and merge. Thanks for the confirmation.

hargup commented:
> (Y) nice work

@prudhvid you can use emoji `:+1:` will give :+1: http://www.emoji-cheat-sheet.com/

nevinvalsaraj commented:
@icyflame  haha :+1:

icyflame commented:
@nevinvalsaraj @prudhvid Thanks a lot for all the feedback. I am merging this into the branch `indigo-devel`.
